managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -780,12 +1088,99 @@ private boolean beforeAddEntry(OpAddEntry addOperation) {
+            entry.release();
+            switch (offerEntryResult) {
+                case SUCCESS:
+                    //happy case
+                    return;
+                case FAIL_SEGMENT_CLOSED:
+                    log.debug("segment closed");
+                    return;
+                case FAIL_BUFFER_FULL:
+                    log.debug("buffer full");
+                    break;
+            }
+        }
+
+        if (offloadEntryFillTask == null || offloadEntryFillTask.isDone()) {
+            offloadEntryFillTask = new CompletableFuture<>();
+            scheduledExecutor.schedule(safeRun(() -> entryFillLoop(currentOffloadHandle, positionNextToOffered,
+                    PositionImpl.get(offeringPosition),
+                    offloadEntryFillTask)), 100, TimeUnit.MILLISECONDS);
+        } // else fill when next entry added
+    }
+
+    private void entryFillLoop(OffloadHandle offloadHandle,
+                               PositionImpl beginPosition, PositionImpl endPosition,
+                               CompletableFuture<Void> offloadEntryFillTask) {
+        asyncReadEntry(beginPosition, new ReadEntryCallback() {
+            void delayExecute(OffloadHandle OffloadHandle,
+                              PositionImpl beginPosition, PositionImpl endPosition,
+                              CompletableFuture<Void> offloadEntryFillTask) {
+                scheduledExecutor
+                        .schedule(() -> entryFillLoop(OffloadHandle, beginPosition, endPosition,
+                                offloadEntryFillTask)
+                                , 100, TimeUnit.MILLISECONDS);
+            }
+
+            @Override
+            public void readEntryComplete(Entry entry, Object ctx) {
+                final OfferEntryResult offerEntryResult = offloadHandle.offerEntry(entry);
+                entry.release();
+
+                switch (offerEntryResult) {
+                    case FAIL_BUFFER_FULL:
+                        delayExecute(offloadHandle, beginPosition, endPosition, offloadEntryFillTask);
+                        break;
+                    case FAIL_SEGMENT_CLOSED:
+                        log.debug("segment closed");
+                        break;
+                    case SUCCESS:
+                        if (beginPosition == endPosition) {
+                            offloadEntryFillTask.complete(null);
+                        } else {
+                            final PositionImpl nextPos = getNextValidPosition(beginPosition);
+                            entryFillLoop(offloadHandle, nextPos, endPosition,
+                                    offloadEntryFillTask);
+                        }
+                        break;
+                }
+            }
+
+            @Override
+            public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
+                offloadEntryFillTask.completeExceptionally(exception);
```

Comment:
I don't see anyone handle the error from the offloadEntryFillTask, so we might need to log errors?